### PR TITLE
Basic component refactor

### PR DESCRIPTION
**Pull request by Darktex**
This PR finally breaks up the big script into smaller components.

This PR doesn't make _any_ other changes to names, making this into an actual torchtune recipe and so on. This is left for future PRs to stack on top.

# NOT YET WORKING AFTER REBASE
We get this error: P1780237599

Tested:

Screenshot 2025-04-08 at 3 02 52 PM

### CODE REVIEW

File: torchtune/dev/rl/workers/parameter_servers/vllm.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        if self.vllm_weight_versions[worker_id] == self.version:
+            log.info(
+                f"skipping update for {worker_id=}, {self.version=}, {self.vllm_weight_versions[worker_id]=}"
+            )
+            return True
+        return False
+
+    def _init_model_update_group(self, worker_id):
+        # here again, I want to grab the tp size from the vLLM worker... :(
+        # llm.llm_engine.parallel_config.tensor_parallel_size
```
Comment (mikaylagawarecki) [inline torchtune/dev/rl/workers/parameter_servers/vllm.py:93]: Hmmm I think #33 might have been lost in the refactor, I can def fix this after this PR lands though!!

Comment (Darktex) [reply torchtune/dev/rl/workers/parameter_servers/vllm.py:93]: No worries, that was before rebase! Let me update now. It's not working tho

File: torchtune/dev/rl/workers/trainers/pytorch_actor.py [hunk 2]
```diff
@@ -0,0 +1,797 @@
+import functools
```
[the remark this answers is not in the document]
Comment (Darktex) [reply torchtune/dev/rl/workers/trainers/pytorch_actor.py:1]: This is a good point, noted. Let me think about this and see what we can do.

One of the things that I would like to point out is that in online RL we are dealing with one more level of encapsulation than "normal" training: ie we have the normal training as usual, but ALSO on top of it we have a coordination layer between the datacollector & trainer. I do think that it is valuable to keep them separate, because different people will interact with different parts of the codebase.

The coordination part will be touched by more infra people who want to implement things like fault tolerance, more asynchronicity etc, while the ML people will want to touch primarily the training loop: for example, I'm thinking someone who wants to switch from PPO to GRPO (or whatever on-policy RL algo is trendy).

So, I was thinking that now we have a 2-dimensional recipe where Actors are really the first dimension, and we have a coordination recipe sit on top of it.

How could we structure this better?

File: torchtune/dev/rl/workers/datacollectors/sync.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
+import time
```
[the remark this answers is not in the document]
Comment (vmoens) [reply torchtune/dev/rl/workers/datacollectors/sync.py:1]: I can see that point and I also understand that a large chunk of the user base wants to see single file implementations to navigate the code with ease. 
I think the challenge we want to tackle (at least on the RL / core side) is that we want to give tools for people to do RL training at scale.
We want to operate over multiple modalities:

- losses (SPO, GRPO, the other one from Yoshua Bengio, ...)
- scale (3B, 7B, 70B)
- model (qwen, llama)
- task (datasets, agents)
- dist backend (ray, monarch)
- (perhaps inference backend: vllm or other?)

The only way to scale a single script in my opinion is to have each of these things in separate files and compose them (losses hosted somewhere in the repo, single / multi-node collectors + updaters somewhere else, ...)

I have the impression that tune has so far operated under the assumption that the items I listed above were enough to motivate a duplication of the script (correct me if I'm wrong - not sure what the cutoff is), but the cartesian product of these all is growing rapidly. My view would be to have a clear dimension along which we think the script needs to be re-written and others along which we're ok having a single entry point. Once that is decided I think it'll be clearer what component should be hosted in the executable script and what should be hosted in tune lib (bc it's used across scripts). Does that make sense?

[the remark this answers is not in the document]
Comment (pbontrager) [reply torchtune/dev/rl/workers/datacollectors/sync.py:1]: These are all good points and highlight the difficulty of creating and maintaining a library like this. But this is also where the incredible value comes to the user. In the future it might be possible again to have nice abstractions that handle lots of use cases, but in the current era of rapid change and need for extreme optimization, having an offering that 
1. shows a single current "right" way to do things
2. exposes most of the steps in training so users can understand what's happening and debug
3. avoids as much abstraction as possible so there is minimal friction for users to try new, unanticipated, ideas
is  very important and what we've tried to build with torchtune.

Given that, my comments would be:
- We don't want a single script to support multiple backends (ray, vllm, etc). If we want to host both we'd have two recipes but more likely we'd replace one with another as you can't optimize two backends well.
- Scale I believe can be handled in at most two recipes. 1 for distributed that's what we're building now and is ultra async, I'm not aware of any issues so far that requires a split approach. (the other would be ultra efficient single device but that's not a priority now)
- Currently models and their distributed strategies are coupled and orthogonal to the recipe. So they should handle the specific meta information needed for different scale models.
- Datasets are currently  also decoupled from the recipe and this approach can be extended to environments/tasks/rewards
- losses should be taken case by case. If the loss logic can nicely all be coupled together then it's natural to have it in a loss module. If it requires lots of unique changes then it should split off to be its own recipe. In general the idea is to be able to quickly spin a new "thorny" algorithm into its own recipe and then merge them back once the abstraction is clearer (we should merge DPO with SFT for example)

As a final comment, this very explicit approach has costs that we're typically taught as SWE's to avoid, and that is in code duplication and extra maintenance burden. But we believe this is what's needed in the current research climate and it has seemed to resonate based on the number of libraries coming out based on a similar concept.

The recipe should expose a lot of the orchestrator and and the explicit steps that is happening that constitutes training and generation. A user should be able to read through each actor class in the recipe file and know what's happening at each step and know exactly where they'd place their own code to customize (minimal magic behind the scenes) and where we do use utilites, ideally users don't have to traverse more than one level down to see exactly what's happening.

It's hard to precisely define what to expose and what to hide. I'd be happy to take a stab at co-designing this refactor to help make it more concrete.
